File: 975-range-sum-of-bst/range-sum-of-bst.py
# class Solution:
#     def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
#         if not root:
#             return 0
#         sum = 0
#         if root.val > L:
#             sum += self.rangeSumBST(root.left, L, R)
#         if root.val < R:
#             sum += self.rangeSumBST(root.right, L, R)
#         if L <= root.val <= R:
#             sum += root.val     
#         return sum
class Solution:
    def lvlorder(self, root):
        res = []
        q = collections.deque()
        q.append(root)
        while q:
            lvl = []
            for i in range(len(q)):
                node = q.popleft()
                lvl.append(node)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(lvl)
        return res

    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if not root:
            return 0
        
        res = 0
        q = collections.deque()
        q.append(root)

        while q:
            node = q.popleft()
            if node.val >= low and node.val <= high:
                res += node.val
            if node.left and node.val > low:
                q.append(node.left)
            if node.right and node.val < high:
                q.append(node.right)
        
        return res


# class Solution:
#     def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
#         if root:
#             if root.val<low:
#                 return self.rangeSumBST(root.right,low,high)
#             elif root.val>high:
#                 return self.rangeSumBST(root.left,low,high)
#             return root.val + self.rangeSumBST(root.left,low,high) + self.rangeSumBST(root.right,low,high)
#         else:
#             return 0
# class Solution:
#     def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
#         if root == None: return 0
#         if root.val > R: return self.rangeSumBST(root.left,L,R)
#         if root.val < L: return self.rangeSumBST(root.right,L,R)
#         return root.val + self.rangeSumBST(root.left,L,R) + self.rangeSumBST(root.right,L,R)      
                   

File: 975-range-sum-of-bst/range-sum-of-bst.cpp
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        if(!root) return 0;
        int sum=0;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            TreeNode* node=q.front();
            q.pop();
            if (node->val >= low && node->val <= high) {
                sum += node->val;
            }
            if (node->left && node->val > low) {
                q.push(node->left);
            }
            if (node->right && node->val < high) {
                q.push(node->right);
            }
        }
        return sum;
        
    }
};
 
// class Solution {
// public:
//     int rangeSumBST(TreeNode* root, int low, int high) {
//         if(root==NULL) return 0;
//         if(root->val < low) return rangeSumBST(root->right,low,high);
//         if(root->val > high) return rangeSumBST(root->left,low,high);
//         return root->val+rangeSumBST(root->left,low,high)+rangeSumBST(root->right,low,high);
//     }
// };